src/mam4xx/ndrop.hpp

```diff
@@ -22,87 +22,110 @@ class NDrop {
-                        (ekkp(k) * (qold(kp1) - qold(k) +
-                                    (qactold(kp1) * (1 - overlapp(k))))) +
-                        (ekkm(k) * (qold(km1) - qold(k) +
-                                    (qactold(km1) * (1 - overlapm(k)))))));
-        } else {
-          q(k) = qold(k) +
-                 (dt *
-                  (src(k) + (ekkp(k) * ((overlapp(k) * qold(kp1)) - qold(k))) +
-                   (ekkm(k) * ((overlapm(k) * qold(k)) - qold(k)))));
-        }
-        // force to non-negative
-        q(k) = max(q(k), 0);
-      });
-}
+  // the qactold*(1-overlap) terms are resuspension of activated material
+
+  if (is_unact) {
+    q = qold_k +
+        (dt *
+         (-src +
+          (ek_kp1 * (qold_kp1 - qold_k + (qactold_kp1 * (1 - overlap_kp1)))) +
+          (ek_km1 * (qold_km1 - qold_k + (qactold_km1 * (1 - overlap_km1))))));
+  } else {
+    q = qold_k + (dt * (src + (ek_kp1 * ((overlap_kp1 * qold_kp1) - qold_k)) +
+                        (ek_km1 * ((overlap_km1 * qold_k) - qold_k))));
+  }
+  // force to non-negative
+  q = max(q, 0);
+} // end explmix
+
+// calculates maximum supersaturation for multiple
+// competing aerosol modes.
+// Abdul-Razzak and Ghan, A parameterization of aerosol activation.
+// 2. Multiple aerosol types. J. Geophys. Res., 105, 6837-6844.
+KOKKOS_INLINE_FUNCTION
+void maxsat(const Real zeta,   // [dimensionless]
+            const Real eta[4], // [dimensionless]
+            const Real nmode,  // modes
+            const Real smc[4], // critical supersaturation for number mode
+                               // radius [fraction]
+            Real &smax         // maximum supersaturation [fraction] (output)
+) {
+  // abdul-razzak functions of width
+  Real f1[AeroConfig::num_modes()];
+  Real f2[AeroConfig::num_modes()];
+
+  Real sum = 0;
+  Real g1, g2;
+  bool weak_forcing = true; // whether forcing is sufficiently weak or not
+
+  for (int m = 0; m < nmode; m++) {
+    if (zeta > 1e5 /*FIXME: BAD CONSTANT*/ * eta[m] ||
+        smc[m] * smc[m] > 1e5 /*FIXME: BAD CONSTANT*/ * eta[m]) {
+      // weak forcing. essentially none activated
+      smax = 1e-20; /*FIXME: BAD CONSTANT*/
+    } else {
+      // significant activation of this mode. calc activation of all modes.
+      weak_forcing = false;
+      break;
+    }
+  }
+
+  if (weak_forcing)
+    return;
+
+  for (int m = 0; m < nmode; m++) {
+    f1[m] = 0.5 * haero::exp(2.5 * haero::log(modes(m).mean_std_dev) *
+                             haero::log(modes(m).mean_std_dev));
+    f2[m] = 1.0 + 0.25 * haero::log(modes(m).mean_std_dev);
+    if (eta[m] > 1e-20 /*FIXME: BAD CONSTANT*/) {
+      g1 = (zeta / eta[m]) * haero::sqrt(zeta / eta[m]);
+      g2 = (smc[m] / haero::sqrt(eta[m] + 3.0 * zeta)) *
+           haero::sqrt(smc[m] / haero::sqrt(eta[m] + 3.0 * zeta));
+      sum += (f1[m] * g1 + f2[m] * g2) / (smc[m] * smc[m]);
+    } else {
+      sum = 1e20; /*FIXME: BAD CONSTANT*/
```

Comment:
`sum=1e20 `-> `sum = 1./small`;

Reply:
Or we could even have another constant `const Real big = 1/small;`